README.md

```diff
@@ -1,16 +1,1831 @@
+type Promise struct {
+  v       SemVer
+
+  run     &Task
+  # Receipt of the invocation that caused this invocation
+  cause   optional &Invocation
+  # Task authorization. If omitted can be interpreted as requires
+  # authorization 
+  auth    optional &Authorization
+
+  meta    {String : any}
+
+  prf     [&UCAN]
+}
+
+# Promise is a way to reference result of the invocation
+type Await union {
+  | &Promise    "await/*"
+  | &Promise    "await/ok"
+  | &Promise    "await/error"
+} representation keyed
+```
+
+# 3 Task
+
+A Task is the smallest unit of work that can be requested from a UCAN. It describes one `(resource, ability, input)` triple. The `input` field is free form, and depend on the specific resource and ability being interacted with, and is not described in this specification.
+
+Using the JavaScript analogy from the introduction, a Task is similar to wrapping a call in an anonymous function:
+
+```json
+{
+  "do": "msg/send",
+  "with": "mailto:alice@example.com",
+  "input": {
+    "to": [
+      "bob@example.com",
+      "carol@example.com"
+    ],
+    "subject": "hello",
+    "body": "world"
+  }
+}
+```
+
+```js
+// Pseudocode
+() =>
+  msg.send("mailto:alice@example.com", {
+    to: ["bob@example.com", "carol@example.com"],
+    subject: "hello",
+    body: "world"
+  })
+```
+
+Later, when we explore promise [pipelines], this also includes capturing the promise:
+
+```json
+{
+  "bafyreid32aunjg6g6buib7bva2sa5ni4c6eaftgykyh6z6rnhcjxh4yk2y": {
+    "do": "crud/read",
+    "with": "https://exmaple.com/mailinglist"
+  },
+  "bafyreigfusg7tgegda7pxdn7px5vs7wxqk5nybr5ewjht4xas357x6ryia": {
+    "do": "msg/send",
+    "with": "mailto://alice@example.com",
+    "input": {
+      "to": {
+        "await/ok": {
+          "/": "bafyreifkeu34pzda27fozfcboh25psekhqehkvnekltoiaspb5jsp6pj5q"
+        }
+      },
+      "subject": "hello",
+      "body": "world"
+    }
+  }
+}
+```
+
+```js
+// Pseudocode
+const mailingList = crud.read("https://exmaple.com/mailinglist");
+const sendEmail = msg.send("mailto://alice@example.com", {
+  to: mailingList.await().ok,
+  subject: "hello",
+  body: "world"
+});
+```
+
+## 3.1 Schema
+
+```ipldsch
+type Task struct {
+  with    URI
+  do      Ability
+  input   {String: any}
+
+  nnc     string
+}
+
+
+type URI string
+type Ability string
+
+```
+
+## 3.2 Fields
+
+### 3.2.1 UCAN Task Version
+
+The `v` field MUST contain the SemVer-formatted version of the UCAN Task Specification that this struct conforms to.
+
+### 3.2.2 Resource
+
+The `with` field MUST contain the [URI](https://en.wikipedia.org/wiki/Uniform_Resource_Identifier) of the resource being accessed. If the resource being accessed is some static data, it is RECOMMENDED to reference it by the [`data`](https://en.wikipedia.org/wiki/Data_URI_scheme), [`ipfs`](https://docs.ipfs.tech/how-to/address-ipfs-on-web/#native-urls), or [`magnet`](https://en.wikipedia.org/wiki/Magnet_URI_scheme) URI schemes.
```

Comment:
Some early feedback that I got from implementers of my PR is that `with` is often a keyword in a lot of languages, which means they need to use workarounds like with_`.

Any thoughts on changing the label to something like `using` or `on` (or something better)? I'm also happy to leave it as-is until someone suggests a better term.

Reply:
> Some early feedback that I got from implementers of my PR is that with is often a keyword in a lot of languages, which means they need to use workarounds like with_.
>
> Any thoughts on changing the label to something like using or on (or something better)? I'm also happy to leave it as-is until someone suggests a better term.

We also despise `with` in UCANs for same reason (`with` is a keyword in JS). I'd be delighted to get rid of `with`, +1 for `on`, otherwise simply `uri` would do too.

Reply:
Either way I would much rather do the name tweaking after we settle everything else, so we can re-generating all the datasets once we settle all the rest.

Reply:
Works for me 👍